src/test/auxiliary/issue_8401.rs

```diff
@@ -22,5 +22,5 @@ fn bar<T>(_: &mut A, _: &T) {}
 
 fn foo<T>(t: &T) {
     let b = B;
-    bar(unsafe { mem::transmute(&b as &A) }, t)
+    bar(&mut b as &mut A, t)
```

Comment:
Did you run this test? I think that the binding to `b` needs to be `mut` for this to work? (I'd recommend being sure to run the tests for these cases).